// ngx-ui-carousel/esm2015/lib/directives/ui-lazy-load.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, Renderer2 } from '@angular/core';
export class UILazyloadDirective {
    /**
     * @param {?} el
     * @param {?} renderer
     */
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    load() {
        /** @type {?} */
        const img = this.el.nativeElement;
        if (img.src) {
            return;
        }
        img.src = this.imgUrl;
    }
}
UILazyloadDirective.decorators = [
    { type: Directive, args: [{
                selector: '[libUiLazyLoad]',
                exportAs: 'ui-lazy-load'
            },] }
];
/** @nocollapse */
UILazyloadDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
UILazyloadDirective.propDecorators = {
    imgUrl: [{ type: Input, args: ['imgUrl',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    UILazyloadDirective.prototype.imgUrl;
    /**
     * @type {?}
     * @private
     */
    UILazyloadDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    UILazyloadDirective.prototype.renderer;
}
