// ngx-ui-carousel/esm2015/lib/color-config.class.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class UiCarouselColorConfig {
    /**
     * @param {?=} dotsColor
     * @param {?=} arrowColor
     * @param {?=} itemBackgroundColor
     * @param {?=} printButtonBackgroundColor
     * @param {?=} printButtonFontColor
     */
    constructor(dotsColor, arrowColor, itemBackgroundColor, printButtonBackgroundColor, printButtonFontColor) {
        this.dotsColor = dotsColor ? dotsColor : '#F5F5F5';
        this.arrowColor = arrowColor ? arrowColor : '#F5F5F5';
        this.itemBackgroundColor = itemBackgroundColor ? itemBackgroundColor : '#CCCCCC';
        this.printButtonBackgroundColor = printButtonBackgroundColor ? printButtonBackgroundColor : '#3d3d3d';
        this.printButtonFontColor = printButtonFontColor ? printButtonFontColor : '#FFFFFF';
    }
}
if (false) {
    /** @type {?} */
    UiCarouselColorConfig.prototype.dotsColor;
    /** @type {?} */
    UiCarouselColorConfig.prototype.arrowColor;
    /** @type {?} */
    UiCarouselColorConfig.prototype.itemBackgroundColor;
    /** @type {?} */
    UiCarouselColorConfig.prototype.printButtonBackgroundColor;
    /** @type {?} */
    UiCarouselColorConfig.prototype.printButtonFontColor;
}
