// ngx-ui-carousel/esm5/lib/components/arrow.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef, Renderer2 } from '@angular/core';
import { UiCarouselColorConfig } from '../color-config.class';
var ArrowComponent = /** @class */ (function () {
    function ArrowComponent(_el, _renderer2) {
        this._el = _el;
        this._renderer2 = _renderer2;
        this.disabled = true;
        this.onArrowClick = new EventEmitter();
    }
    Object.defineProperty(ArrowComponent.prototype, "colorConfig", {
        get: /**
         * @return {?}
         */
        function () {
            return this._colorConfig;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._colorConfig = value;
            this.createStyle();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ArrowComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this._colorConfig) {
            this.colorConfig = new UiCarouselColorConfig();
        }
    };
    /**
     * @return {?}
     */
    ArrowComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (!this.disabled) {
            this.onArrowClick.emit();
            this.createStyle();
        }
    };
    /**
     * @return {?}
     */
    ArrowComponent.prototype.createStyle = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var styling = "\n      .arrow::before, .arrow::after {\n        background: " + this.colorConfig.arrowColor + ";\n      }\n      ";
        if (this.styleElement) {
            this._renderer2.removeChild(this._renderer2.parentNode(this.styleElement), this.styleElement);
        }
        this.styleElement = this._renderer2.createElement('style');
        this.styleElement.appendChild(this._renderer2.createText(styling));
        this._el.nativeElement.appendChild(this.styleElement);
    };
    ArrowComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'arrow',
                    template: "\n    <div #arrow class=\"arrow\" (click)=\"onClick()\"\n      [ngClass]=\"{'left': dir === 'left', 'right': dir === 'right', 'disabled': disabled}\">\n      <style #shadow-styling [innerHTML]=\"styling\">\n      </style>\n      </div>\n  ",
                    styles: ["\n  .arrow {\n    position: absolute;\n    height: 50px;\n    width: 30px;\n    opacity: .6;\n    user-select: none;\n    -moz-user-select: none;\n    -khtml-user-select: none;\n    -webkit-user-select: none;\n    -o-user-select: none;\n    z-index: 1000;\n  }\n\n  .arrow.right {\n    right: 15px;\n    top: 50%;\n\n    transform: scaleX(-1) translateY(-50%);\n    -moz-transform: scaleX(-1) translateY(-50%);\n    -o-transform: scaleX(-1) translateY(-50%);\n    -webkit-transform: scaleX(-1) translateY(-50%);\n    -ms-transform: scaleX(-1) translateY(-50%);\n    filter: FlipH;\n    -ms-filter: \"FlipH\";\n  }\n\n  .arrow.left{\n    left: 15px;\n    top: 50%;\n    transform: translateY(-50%);\n    -moz-transform: translateY(-50%);\n    -webkit-transform: translateY(-50%);\n    -o-transform: translateY(-50%);\n    -ms-transform: translateY(-50%);\n  }\n\n  .arrow:hover{\n    opacity: .8;\n    cursor: pointer;\n  }\n\n  .arrow:before{\n    content: \"\";\n    height: 3px;\n    width: 30px;\n    display: block;\n    position: absolute;\n    top: 14px;\n    transform: rotate(-45deg);\n    -moz-transform: rotate(-45deg);\n    -webkit-transform: rotate(-45deg);\n    -o-transform: rotate(-45deg);\n    -ms-transform: rotate(-45deg);\n  }\n  .arrow:after{\n    content: \"\";\n    height: 3px;\n    width: 30px;\n    display: block;\n    transform: rotate(45deg);\n    -moz-transform: rotate(45deg);\n    -webkit-transform: rotate(45deg);\n    -o-transform: rotate(45deg);\n    -ms-transform: rotate(45deg);\n    position: absolute;\n    bottom: 14px;\n  }\n  .arrow.disabled{\n    opacity: .4;\n  }\n  .arrow.disabled:hover{\n    opacity: .4;\n    cursor: pointer;\n  }"]
                }] }
    ];
    /** @nocollapse */
    ArrowComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    ArrowComponent.propDecorators = {
        dir: [{ type: Input }],
        disabled: [{ type: Input }],
        colorConfig: [{ type: Input, args: ['colorConfig',] }],
        onArrowClick: [{ type: Output }]
    };
    return ArrowComponent;
}());
export { ArrowComponent };
if (false) {
    /** @type {?} */
    ArrowComponent.prototype.dir;
    /** @type {?} */
    ArrowComponent.prototype.disabled;
    /** @type {?} */
    ArrowComponent.prototype.onArrowClick;
    /** @type {?} */
    ArrowComponent.prototype.styling;
    /** @type {?} */
    ArrowComponent.prototype.styleElement;
    /**
     * @type {?}
     * @private
     */
    ArrowComponent.prototype._colorConfig;
    /**
     * @type {?}
     * @private
     */
    ArrowComponent.prototype._el;
    /**
     * @type {?}
     * @private
     */
    ArrowComponent.prototype._renderer2;
}
