// ngx-ui-carousel/esm5/lib/components/dots.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, HostBinding } from '@angular/core';
import { UiCarouselColorConfig } from '../color-config.class';
var DotsComponent = /** @class */ (function () {
    function DotsComponent() {
        this.activeDot = 0;
        this.colorConfig = new UiCarouselColorConfig();
        this.position = 'left';
        this.onClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    DotsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.numbers = Array(this.dotsCount).fill(0).map((/**
         * @param {?} x
         * @param {?} i
         * @return {?}
         */
        function (x, i) { return i; }));
    };
    /**
     * @param {?} index
     * @return {?}
     */
    DotsComponent.prototype.click = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.onClick.emit(index);
        this.activeDot = index;
    };
    DotsComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'dots',
                    template: "\n    <div class=\"dot\" [ngStyle]=\"{'background-color': colorConfig.dotsColor}\"\n      *ngFor=\"let number of numbers\"\n      (click)=\"click(number)\"\n      [class.active]=\"activeDot === number\">\n    </div>\n  ",
                    styles: ["\n      :host{\n          position: absolute;\n          display: inline-block;\n          z-index: 1000;\n      }\n\n      :host(.left){\n          bottom: 10px;\n          left: 10px;\n      }\n\n      :host(.right){\n          bottom: 10px;\n          right: 10px;\n      }\n\n      :host(.middle){\n          bottom: 20px;\n          left: 50%;\n          transform: translateX(-50%);\n          -webkit-transform: translateX(-50%);\n          -moz-transform: translateX(-50%);\n          -o-transform: translateX(-50%);\n          -ms-transform: translateX(-50%);\n      }\n\n      .dot{\n          height: 10px;\n          width: 10px;\n          border-radius: 5px;\n          background: white;\n          opacity: .5;\n          margin: 0 4px;\n          display: inline-block;\n      }\n\n      .dot:hover{\n          opacity: .9;\n          cursor: pointer;\n      }\n\n      .dot.active{\n          opacity: .9;\n      }\n  "]
                }] }
    ];
    /** @nocollapse */
    DotsComponent.ctorParameters = function () { return []; };
    DotsComponent.propDecorators = {
        activeDot: [{ type: Input }],
        dotsCount: [{ type: Input }],
        colorConfig: [{ type: Input }],
        position: [{ type: HostBinding, args: ['class',] }, { type: Input }],
        onClick: [{ type: Output }]
    };
    return DotsComponent;
}());
export { DotsComponent };
if (false) {
    /** @type {?} */
    DotsComponent.prototype.activeDot;
    /** @type {?} */
    DotsComponent.prototype.dotsCount;
    /** @type {?} */
    DotsComponent.prototype.colorConfig;
    /** @type {?} */
    DotsComponent.prototype.position;
    /** @type {?} */
    DotsComponent.prototype.onClick;
    /** @type {?} */
    DotsComponent.prototype.numbers;
}
