// ngx-ui-carousel/esm5/lib/components/print.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef, Renderer2 } from '@angular/core';
import { UiCarouselColorConfig } from '../color-config.class';
var PrintButtonComponent = /** @class */ (function () {
    function PrintButtonComponent(_el, _renderer2) {
        this._el = _el;
        this._renderer2 = _renderer2;
        this.disabled = true;
        this.colorConfig = new UiCarouselColorConfig();
        this.onButtonClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    PrintButtonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PrintButtonComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (!this.disabled) {
            this.onButtonClick.emit();
        }
    };
    PrintButtonComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'print-button',
                    template: "\n    <div #printButton\n    class=\"print-button\"\n    (click)=\"onClick()\"\n    [ngClass]=\"{'disabled': disabled}\"\n    [ngStyle]=\"{'background-color': colorConfig.printButtonBackgroundColor}\">\n      <span [ngStyle]=\"{'color': colorConfig.printButtonFontColor}\">print</span>\n    </div>\n  ",
                    styles: ["\n  .print-button {\n    position: absolute;\n    height: 30px;\n    line-height: 30px;\n    text-align: center;\n    border-radius: 5px;\n    width: 60px;\n    opacity: .6;\n    user-select: none;\n    -moz-user-select: none;\n    -khtml-user-select: none;\n    -webkit-user-select: none;\n    -o-user-select: none;\n    z-index: 1000;\n    right: 20px;\n    top: 20px;\n  }\n\n  .print-button.disabled{\n    opacity: .4;\n  }\n  .print-button:hover{\n    opacity: .9;\n    cursor: pointer;\n  }\n  .print-button.disabled:hover{\n    opacity: .4;\n    cursor: pointer;\n  }"]
                }] }
    ];
    /** @nocollapse */
    PrintButtonComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    PrintButtonComponent.propDecorators = {
        dir: [{ type: Input }],
        disabled: [{ type: Input }],
        colorConfig: [{ type: Input }],
        onButtonClick: [{ type: Output }]
    };
    return PrintButtonComponent;
}());
export { PrintButtonComponent };
if (false) {
    /** @type {?} */
    PrintButtonComponent.prototype.dir;
    /** @type {?} */
    PrintButtonComponent.prototype.disabled;
    /** @type {?} */
    PrintButtonComponent.prototype.colorConfig;
    /** @type {?} */
    PrintButtonComponent.prototype.onButtonClick;
    /** @type {?} */
    PrintButtonComponent.prototype.styling;
    /** @type {?} */
    PrintButtonComponent.prototype.styleElement;
    /**
     * @type {?}
     * @private
     */
    PrintButtonComponent.prototype._el;
    /**
     * @type {?}
     * @private
     */
    PrintButtonComponent.prototype._renderer2;
}
