// ngx-ui-carousel/esm5/lib/components/ui-carousel-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, ElementRef, QueryList, Renderer2, ChangeDetectorRef, ContentChild, Input } from '@angular/core';
import { UILazyloadDirective } from '../directives/ui-lazy-load.directive';
import { UiCarouselColorConfig } from '../color-config.class';
var UiCarouselItemComponent = /** @class */ (function () {
    function UiCarouselItemComponent(renderer, ref) {
        this.renderer = renderer;
        this.ref = ref;
        this.currentPosition = 0;
        this.position = 0;
        this.colorConfig = new UiCarouselColorConfig();
    }
    /**
     * @param {?} position
     * @return {?}
     */
    UiCarouselItemComponent.prototype.moveTo = /**
     * @param {?} position
     * @return {?}
     */
    function (position) {
        this.renderer.setStyle(this.el.nativeElement, 'transform', 'translate3d(' + position + 'px, 0px, 0px)');
        this.renderer.setStyle(this.el.nativeElement, '-webkit-transform', 'translate3d(' + position + 'px, 0px, 0px)');
        this.renderer.setStyle(this.el.nativeElement, '-moz-transform', 'translate3d(' + position + 'px, 0px, 0px)');
        this.renderer.setStyle(this.el.nativeElement, '-o-transform', 'translate3d(' + position + 'px, 0px, 0px)');
        this.renderer.setStyle(this.el.nativeElement, '-ms-transform', 'translate3d(' + position + 'px, 0px, 0px)');
    };
    /**
     * @param {?} distance
     * @return {?}
     */
    UiCarouselItemComponent.prototype.moveBy = /**
     * @param {?} distance
     * @return {?}
     */
    function (distance) {
        this.renderer.setStyle(this.el.nativeElement, 'transform', 'translate3d(' + distance + 'px, 0px, 0px)');
        this.renderer.setStyle(this.el.nativeElement, '-webkit-transform', 'translate3d(' + distance + 'px, 0px, 0px)');
        this.renderer.setStyle(this.el.nativeElement, '-moz-transform', 'translate3d(' + distance + 'px, 0px, 0px)');
        this.renderer.setStyle(this.el.nativeElement, '-o-transform', 'translate3d(' + distance + 'px, 0px, 0px)');
        this.renderer.setStyle(this.el.nativeElement, '-ms-transform', 'translate3d(' + distance + 'px, 0px, 0px)');
    };
    /**
     * @param {?} zIndex
     * @return {?}
     */
    UiCarouselItemComponent.prototype.setzIndex = /**
     * @param {?} zIndex
     * @return {?}
     */
    function (zIndex) {
        this.renderer.setStyle(this.el.nativeElement, 'z-index', zIndex);
    };
    /**
     * @return {?}
     */
    UiCarouselItemComponent.prototype.disableTransition = /**
     * @return {?}
     */
    function () {
        this.renderer.setStyle(this.el.nativeElement, 'transition', 'none');
        this.renderer.setStyle(this.el.nativeElement, '-moz-transition', 'none');
        this.renderer.setStyle(this.el.nativeElement, '-webkit-transition', 'none');
        this.renderer.setStyle(this.el.nativeElement, '-o-transition', 'none');
        this.renderer.setStyle(this.el.nativeElement, '-ms-transition', 'none');
    };
    /**
     * @return {?}
     */
    UiCarouselItemComponent.prototype.enableTransition = /**
     * @return {?}
     */
    function () {
        this.renderer.setStyle(this.el.nativeElement, 'transition', 'transform');
        this.renderer.setStyle(this.el.nativeElement, '-moz-transition', 'transform');
        this.renderer.setStyle(this.el.nativeElement, '-webkit-transition', 'transform');
        this.renderer.setStyle(this.el.nativeElement, '-o-transition', 'transform');
        this.renderer.setStyle(this.el.nativeElement, '-ms-transition', 'transform');
        this.renderer.setStyle(this.el.nativeElement, 'transition-duration', this.speed + 'ms');
        this.renderer.setStyle(this.el.nativeElement, '-moz-transition-duration', this.speed + 'ms');
        this.renderer.setStyle(this.el.nativeElement, '-webkit-transition-duration', this.speed + 'ms');
        this.renderer.setStyle(this.el.nativeElement, '-o-transition-duration', this.speed + 'ms');
        this.renderer.setStyle(this.el.nativeElement, '-ms-transition-duration', this.speed + 'ms');
    };
    /**
     * @param {?} duration
     * @return {?}
     */
    UiCarouselItemComponent.prototype.fadeOut = /**
     * @param {?} duration
     * @return {?}
     */
    function (duration) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            _this.renderer.setStyle(_this.el.nativeElement, 'opacity', '0');
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.renderer.setStyle(_this.el.nativeElement, 'opacity', '1');
                resolve();
            }), duration);
        }));
    };
    /**
     * @param {?} duration
     * @return {?}
     */
    UiCarouselItemComponent.prototype.fadeIn = /**
     * @param {?} duration
     * @return {?}
     */
    function (duration) {
        var _this = this;
        this.renderer.setStyle(this.el.nativeElement, 'opacity', '0');
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.renderer.setStyle(_this.el.nativeElement, 'transition', 'opacity ' + duration + 'ms');
            _this.renderer.setStyle(_this.el.nativeElement, 'opacity', '1');
        }), 0);
    };
    /**
     * @return {?}
     */
    UiCarouselItemComponent.prototype.lazyLoad = /**
     * @return {?}
     */
    function () {
        this.lazyLoadedImages.forEach((/**
         * @param {?} img
         * @return {?}
         */
        function (img) {
            img.load();
        }));
    };
    UiCarouselItemComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'ngx-ui-carousel-item',
                    template: "\n    <div #carouselItem\n        [ngStyle]=\"{'background-color': backgroundColor ? backgroundColor : colorConfig.itemBackgroundColor}\"\n        class=\"ui-carousel-item fade\" tabindex=\"-1\"\n        style=\"outline: none\">\n      <ng-content></ng-content>\n    </div>\n  ",
                    styles: ["\n  :host {\n    display:block!important;\n  }\n\n  .ui-carousel-item{\n    user-select: none;\n    -moz-user-select: none;\n    -khtml-user-select: none;\n    -webkit-user-select: none;\n    -o-user-select: none;\n    -ms-user-select: none;\n  }\n\n  .transition{\n    transition: transform;\n    -moz-transition: transform;\n    -webkit-transition: transform;\n    -o-transition: transform;\n    -ms-transition: transform;\n    transition-timing-function: ease;\n    -moz-transition-timing-function: ease;\n    -o-transition-timing-function: ease;\n    -ms-transition-timing-function: ease;\n  }\n\n  .ui-carousel-item{\n    // height: 100%;\n    width: 100%;\n    position: absolute;\n    overflow: hidden;\n  }\n\n  .fade{\n    opacity: 1;\n    -webkit-transition: opacity .5s ease-in;\n    -moz-transition: opacity .5s ease-in;\n    -ms-transition: opacity .5s ease-in;\n    -o-transition: opacity .5s ease-in;\n    transition: opacity .5s ease-in;\n  }\n\n  .fade-out{\n    opacity: 0;\n  }\n  "]
                }] }
    ];
    /** @nocollapse */
    UiCarouselItemComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ChangeDetectorRef }
    ]; };
    UiCarouselItemComponent.propDecorators = {
        el: [{ type: ViewChild, args: ['carouselItem',] }],
        lazyLoadedImages: [{ type: ContentChild, args: [UILazyloadDirective,] }],
        backgroundColor: [{ type: Input }]
    };
    return UiCarouselItemComponent;
}());
export { UiCarouselItemComponent };
if (false) {
    /** @type {?} */
    UiCarouselItemComponent.prototype.el;
    /** @type {?} */
    UiCarouselItemComponent.prototype.lazyLoadedImages;
    /** @type {?} */
    UiCarouselItemComponent.prototype.backgroundColor;
    /** @type {?} */
    UiCarouselItemComponent.prototype.speed;
    /** @type {?} */
    UiCarouselItemComponent.prototype.currentPosition;
    /** @type {?} */
    UiCarouselItemComponent.prototype.position;
    /** @type {?} */
    UiCarouselItemComponent.prototype.zIndex;
    /** @type {?} */
    UiCarouselItemComponent.prototype.colorConfig;
    /**
     * @type {?}
     * @private
     */
    UiCarouselItemComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    UiCarouselItemComponent.prototype.ref;
}
