// ngx-ui-carousel/esm5/lib/directives/ui-lazy-load.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, Renderer2 } from '@angular/core';
var UILazyloadDirective = /** @class */ (function () {
    function UILazyloadDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    UILazyloadDirective.prototype.load = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var img = this.el.nativeElement;
        if (img.src) {
            return;
        }
        img.src = this.imgUrl;
    };
    UILazyloadDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[libUiLazyLoad]',
                    exportAs: 'ui-lazy-load'
                },] }
    ];
    /** @nocollapse */
    UILazyloadDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    UILazyloadDirective.propDecorators = {
        imgUrl: [{ type: Input, args: ['imgUrl',] }]
    };
    return UILazyloadDirective;
}());
export { UILazyloadDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UILazyloadDirective.prototype.imgUrl;
    /**
     * @type {?}
     * @private
     */
    UILazyloadDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    UILazyloadDirective.prototype.renderer;
}
