// ngx-ui-carousel/esm2015/lib/components/dots.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, HostBinding } from '@angular/core';
import { UiCarouselColorConfig } from '../color-config.class';
export class DotsComponent {
    constructor() {
        this.activeDot = 0;
        this.colorConfig = new UiCarouselColorConfig();
        this.position = 'left';
        this.onClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.numbers = Array(this.dotsCount).fill(0).map((/**
         * @param {?} x
         * @param {?} i
         * @return {?}
         */
        (x, i) => i));
    }
    /**
     * @param {?} index
     * @return {?}
     */
    click(index) {
        this.onClick.emit(index);
        this.activeDot = index;
    }
}
DotsComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'dots',
                template: `
    <div class="dot" [ngStyle]="{'background-color': colorConfig.dotsColor}"
      *ngFor="let number of numbers"
      (click)="click(number)"
      [class.active]="activeDot === number">
    </div>
  `,
                styles: [`
      :host{
          position: absolute;
          display: inline-block;
          z-index: 1000;
      }

      :host(.left){
          bottom: 10px;
          left: 10px;
      }

      :host(.right){
          bottom: 10px;
          right: 10px;
      }

      :host(.middle){
          bottom: 20px;
          left: 50%;
          transform: translateX(-50%);
          -webkit-transform: translateX(-50%);
          -moz-transform: translateX(-50%);
          -o-transform: translateX(-50%);
          -ms-transform: translateX(-50%);
      }

      .dot{
          height: 10px;
          width: 10px;
          border-radius: 5px;
          background: white;
          opacity: .5;
          margin: 0 4px;
          display: inline-block;
      }

      .dot:hover{
          opacity: .9;
          cursor: pointer;
      }

      .dot.active{
          opacity: .9;
      }
  `]
            }] }
];
/** @nocollapse */
DotsComponent.ctorParameters = () => [];
DotsComponent.propDecorators = {
    activeDot: [{ type: Input }],
    dotsCount: [{ type: Input }],
    colorConfig: [{ type: Input }],
    position: [{ type: HostBinding, args: ['class',] }, { type: Input }],
    onClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DotsComponent.prototype.activeDot;
    /** @type {?} */
    DotsComponent.prototype.dotsCount;
    /** @type {?} */
    DotsComponent.prototype.colorConfig;
    /** @type {?} */
    DotsComponent.prototype.position;
    /** @type {?} */
    DotsComponent.prototype.onClick;
    /** @type {?} */
    DotsComponent.prototype.numbers;
}
